File: 2015/c/2015-03.c
#include <stdio.h>
#include <string.h>

#include "tools.h"

int part1(char input[]) {
  const int length = strlen(input);
  int visited[length];
  int pos = 0;
  int nextVisited = 1;

  // Initialize visited
  for (int i = 0; i <= length; i++) visited[i] = 0;

  for (int i = 0; i < length; i++) {
    switch (input[i]) {
      case '<':
        pos--;
        break;

      case '>':
        pos++;
        break;

      case '^':
        pos -= length;
        break;

      case 'v':
        pos += length;
        break;
    }

    int alreadyExists = 0;
    for (int j = 0; j <= length; j++)
      if (visited[j] == pos) alreadyExists = 1;

    if (!alreadyExists) {
      visited[nextVisited] = pos;
      nextVisited++;
    }
  }

  int amountVisited = 1;
  for (int i = 0; i <= length; i++)
    if (visited[i] != 0) amountVisited++;

  return amountVisited;
}

int part2(char input[]) {
  const int length = strlen(input);
  int visited[length];
  int pos[2] = {0, 0};
  int nextVisited = 1;

  // Initialize visited
  for (int i = 0; i <= length; i++) visited[i] = 0;

  for (int i = 0; i < length; i++) {
    int index = i % 2;

    switch (input[i]) {
      case '<':
        pos[index]--;
        break;

      case '>':
        pos[index]++;
        break;

      case '^':
        pos[index] -= length;
        break;

      case 'v':
        pos[index] += length;
        break;
    }

    int alreadyExists = 0;
    for (int j = 0; j <= length; j++)
      if (visited[j] == pos[index]) alreadyExists = 1;

    if (!alreadyExists) {
      visited[nextVisited] = pos[index];
      nextVisited++;
    }
  }

  int amountVisited = 1;
  for (int i = 0; i <= length; i++)
    if (visited[i] != 0) amountVisited++;

  return amountVisited;
}

int main() {
  char *data = getDataFromFile("../data/2015-03.txt");

  printf("Advent of code 2015 #03\n\n");

  printf("----Part 1----\n");
  test("Test 1", part1(">"),          2);
  test("Test 2", part1("^>v<"),       4);
  test("Test 3", part1("^v^v^v^v^v"), 2);
  printf("Answer: %d\n\n", part1(data));

  printf("----Part 2----\n");
  test("Test 1", part2("^v"),          3);
  test("Test 2", part2("^>v<"),        3);
  test("Test 3", part2("^v^v^v^v^v"), 11);
  printf("Answer: %d\n\n", part2(data));

  return 0;
}

File: 2015/js/2015-03.js
const path  = require('path')
const fs    = require('fs')

const data = fs.readFileSync(path.resolve(__dirname, '../data/2015-03.txt'), 'utf8')

function part1(input) {
  let pos = 0
  const visited = [0]

  input.split('').forEach(dir => {
    switch (dir) {
      case '<':
        pos--
        break
      case '>':
        pos++
        break
      case '^':
        pos -= input.length
        break
      case 'v':
        pos += input.length
        break
    }

    if (!visited.includes(pos)) visited.push(pos)
  })

  return visited.length
}

function part2(input) {
  let pos = [0, 0]
  const visited = [0]

  input.split('').forEach((dir, i) => {
    switch (dir) {
      case '<':
        pos[i % 2]--
        break
      case '>':
        pos[i % 2]++
        break
      case '^':
        pos[i % 2] -= input.length
        break
      case 'v':
        pos[i % 2] += input.length
        break
    }

    if (!visited.includes(pos[i % 2])) visited.push(pos[i % 2])
  })

  return visited.length
}

console.clear()

console.log( 'Part 1 tests' )
console.assert( part1('>')          == 2 )
console.assert( part1('^>v<')       == 4 )
console.assert( part1('^v^v^v^v^v') == 2 )

console.log( 'Part 2 tests' )
console.assert( part2('^v')         == 3 )
console.assert( part2('^>v<')       == 3 )
console.assert( part2('^v^v^v^v^v') == 11 )

console.log( `
Advent of code 2015 #03
Part1: ${part1(data)}
Part2: ${part2(data)}
` )